### PR TITLE
Add new debugging system with automatic reverse compatibility

### PR DESCRIPTION
**Pull request by chrisinajar**
Based on https://github.com/AngelArenaAllstars/aaa/pull/80

### CODE REVIEW

File: game/scripts/vscripts/internal/util.lua [hunk 1]
```diff
@@ -6,14 +15,88 @@ function split(s, delimiter)
...
+        table.insert(result, match);
+    end
+    return result;
+end
+
+function TracesFromFilename (filename)
+  local traces = {}
+  local i = 1
+
+  local parts = regexsplit(filename, '%s/\\')
```
Comment (Chronophylos) [inline game/scripts/vscripts/internal/util.lua]: This is not Cross-platform compatible

File: game/scripts/vscripts/internal/util.lua [hunk 2]
```diff
@@ -6,14 +15,88 @@ function split(s, delimiter)
...
+
+-- written by yeahbuddy, taken from https://github.com/AngelArenaAllstars/aaa/pull/80
+-- modified for clarity
+function GetCallingFile (offset)
+  offset = offset or 4
+
+  local str = debug.traceback()
+  local lines = split(str, '\n')
+  local line = lines[offset]
+  local dirName , lineNo= string.match(line, "scripts[/\\]vscripts[/\\](.+).lua:([0-9]+):")
```
Comment (Chronophylos) [inline game/scripts/vscripts/internal/util.lua]: Again this is (if I'm not missing something) not Crossplatform compatible.

File: game/scripts/vscripts/internal/util.lua (hunk 1 above)
Comment (chrisinajar) [inline game/scripts/vscripts/internal/util.lua]: I'm splitting on `%s`, `/`, and `\` here. The regex split method takes in the string and puts it inside a `[^ delim ]` segment, so here it'll grab whitespace or either platform's slash.

File: game/scripts/vscripts/internal/util.lua (hunk 2 above)
Comment (chrisinajar) [inline game/scripts/vscripts/internal/util.lua]: Also splitting on `/` and `\`.

`[/\\]` is a character group of / and \, so it'll grab either.

**Comment by chrisinajar** [comment]
Patterns are confusing because they look like regex's but don't always perform like them. In the code here I use `[/\\]` a lot, which will match either platform's slash since `\` isn't the escape character, `%` is.

**Comment by Chronophylos** [comment]
Sorry I forgot how to regex. :/

**Pull request merged**